candle-examples/examples/llava/main.rs

```diff
@@ -0,0 +1,252 @@
+use candle_transformers::generation::{LogitsProcessor, Sampling};
+use candle_transformers::models::llama::Cache;
+
+use anyhow::{bail, Error as E, Result};
+use candle::{DType, IndexOp, Tensor};
+use candle_nn::VarBuilder;
+use candle_transformers::models::llava::config::{
+    HFGenerationConfig, HFLLaVAConfig, HFPreProcessorConfig,
+};
+use candle_transformers::models::llava::{
+    clip_image_processor::CLIPImageProcessor,
+    config::LLaVAConfig,
+    constants::*,
+    conversation::Conversation,
+    utils::{get_model_name_from_path, process_image, tokenizer_image_token},
+    LLaVA,
+};
+use clap::Parser;
+use hf_hub::api::sync::Api;
+use std::io::Write;
+use std::process::Command;
+use tokenizers::Tokenizer;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about,long_about=None)]
+struct Args {
+    #[arg(long, default_value = "llava-hf/llava-v1.6-vicuna-7b-hf")]
+    model_path: String,
+    #[arg(long)]
+    model_base: Option<String>,
+    #[arg(long, default_value = "candle-examples/examples/llava/llava_logo.png")]
+    image_file: String, // Required
+    #[arg(long)]
+    conv_mode: Option<String>,
+    #[arg(long, default_value_t = 0.2)]
+    temperature: f32,
+    #[arg(long, default_value_t = 512)]
+    max_new_tokens: usize,
+    #[arg(long, action)]
+    load_8bit: bool, // now useless
+    #[arg(long, action)]
+    load_4bit: bool, //now useless
+    #[arg(long, action)]
+    debug: bool, // now useless
+    #[arg(long, action)]
+    cpu: bool,
+    #[arg(long, action)]
+    no_kv_cache: bool,
+    #[arg(long, default_value = "Is this a cat?")]
+    prompt: String,
+    /// The seed to use when generating random samples. Copy from candle llama. Not exist in python llava.
+    #[arg(long, default_value_t = 299792458)]
+    seed: u64,
+}
+
+//from https://github.com/huggingface/candle/blob/main/candle-examples/examples/clip/main.rs
+fn load_image<T: AsRef<std::path::Path>>(
+    path: T,
+    processor: &CLIPImageProcessor,
+    llava_config: &LLaVAConfig,
+    dtype: DType,
+) -> anyhow::Result<((u32, u32), Tensor)> {
+    let img = image::io::Reader::open(path)?.decode()?;
+    let img_tensor = process_image(&img, processor, llava_config)?;
+    Ok(((img.width(), img.height()), img_tensor.to_dtype(dtype)?))
+}
+
+fn get_model_name(path: &str) -> String {
+    path.split('/').last().unwrap().to_string()
+}
+
+fn main() -> Result<()> {
+    let mut args = Args::parse();
+    let device = candle_examples::device(args.cpu)?;
+    println!("Start loading model");
+    let api = Api::new()?;
+    let api = api.model(args.model_path.clone());
+    let model_name = get_model_name(&args.model_path);
+    let (llava_config, tokenizer, clip_vision_config, image_processor) = if model_name
+        .contains("hf")
```

Comment:
Seems a bit sketching to rely on containing `hf`, maybe ends with `-hf` would be more robust? Or maybe it should just be an explicit flag?

Reply:
I modify it to use an explicit --hf flag to determine if it is an HuggingFace model